### PR TITLE
GH-103721: Allow defining GenericAlias-like things

### PR DESCRIPTION
**Pull request by adriangb**
* Issue: gh-103721

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -1336,6 +1336,16 @@ def __dir__(self):
 #   e.g., Dict[T, int].__args__ == (T, int).
 
 
+def _is_special_typing_construct(t) -> bool:
+    return (
+        hasattr(t, "__args__")
+        and
+        hasattr(t, "__origin__")
+        and
+        hasattr(t, "__parameters__")
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
        and hasattr(t, "__origin__")
        and hasattr(t, "__parameters__")
```

File: Lib/typing.py [hunk 2]
```diff
@@ -2469,8 +2479,12 @@ def get_origin(tp):
     """
     if isinstance(tp, _AnnotatedAlias):
         return Annotated
-    if isinstance(tp, (_BaseGenericAlias, GenericAlias,
-                       ParamSpecArgs, ParamSpecKwargs)):
+    if (
+        isinstance(tp, (_BaseGenericAlias, GenericAlias,
+                       ParamSpecArgs, ParamSpecKwargs))
+        or
+        _is_special_typing_construct(tp)
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
        or _is_special_typing_construct(tp)
```

**Review by AlexWaygood — COMMENTED** [review]
Two PEP8 nits:

File: Lib/typing.py [hunk 3]
```diff
@@ -2492,7 +2503,7 @@ def get_args(tp):
     """
     if isinstance(tp, _AnnotatedAlias):
         return (tp.__origin__,) + tp.__metadata__
-    if isinstance(tp, (_GenericAlias, GenericAlias)):
+    if isinstance(tp, (_GenericAlias, GenericAlias)) or _is_special_typing_construct(tp):
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
    if (
        isinstance(tp, (_GenericAlias, GenericAlias))
        or _is_special_typing_construct(tp)
    ):
```

File: Misc/NEWS.d/next/Library/2023-04-23-18-10-10.gh-issue-103721.4Ve-P9.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Allow defining ``GenericAlias``-like things.
+This lets libraries that want to do custom runtime handling of generic parameters and their substitutions integrate with the rest of typing, for example with ``typing.get_args``, ``typing.get_origin``, and as parameters to other generics like ``list``, ``dict`` and so on.
```
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-04-23-18-10-10.gh-issue-103721.4Ve-P9.rst]:
```suggestion
This lets libraries that want to do custom runtime handling of generic
parameters and their substitutions integrate with the rest of typing, for
example with :func:`typing.get_args`, :func:`typing.get_origin`, and as
parameters to other generics like :class:`list`, :class:`dict` and so on.
```

**Review by AlexWaygood — COMMENTED** [review]
Two more small comments below (wrap to 80 chars for `.py` and `.rst` files, use Sphinx markup to link to the rest of the docs).

In general, I think this is a reasonable feature to implement, but I think it should probably be documented somewhere in `typing.rst`.

File: Lib/test/test_typing.py [hunk 5]
```diff
@@ -8606,5 +8606,75 @@ def test_is_not_instance_of_iterable(self):
...
+                # implemented just for easy comparison in the tests
+                # below
+                if not isinstance(other, GenericAliasLike):
+                    return False
+                return (
+                    self.__args__ == other.__args__
+                    and
+                    self.__origin__ == other.__origin__
+                    and
+                    self.__parameters__ == other.__parameters__
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]: oops, more PEP8, sorry

```suggestion
                    and self.__origin__ == other.__origin__
                    and self.__parameters__ == other.__parameters__
```

Comment (adriangb) [reply Lib/test/test_typing.py]: Don't apologize, you're helping me out, I'm just used to living in the "if it doesn't get autoformatted I don't know what to do" world.

**Review by sobolevn — COMMENTED** [review]
I am -1 on this, because it can backfire for a lot of tools that rely on the fact that `{types,typing}.GenericAlias` (and nothing else) represent generic types.

For example, https://github.com/HypothesisWorks/hypothesis/blob/e33377305225df0a4ee50a1bd55ed801df6d1ea2/hypothesis-python/src/hypothesis/strategies/_internal/types.py#L298 (and in many other places).

I think that this needs more visibility and discussion before getting accepted.

**Comment by adriangb** [comment]
Would those things backfire any more now than they currently do? Clearly, this PR doesn't make everything automatically work for every library. Anything that is currently doing introspection and trying to detect `typing.GenericAlias` (which is sort of a private thing, only mentioned once in passing in `typing.rst`) would need to add some sort of `_is_generic_alias_like` check or special case the particular library (Pydantic already has a Hypothesis plugin). But I don't think that makes things worse than the current status quo.

The best alternative to the current implementation/proposal I can think of is to make this stuff more "public" by providing a Protocol, exporting a public `is_generic_alias_like` function, or making `GenericAlias` subclassable. Then things like Hypothesis would "just work", which is I think what you want. I feel that the barrier for that is much higher since `GenericAlias` is this semi-private mostly undocumented thing.

**Comment by JelleZijlstra** [comment]
Let's keep broader discussion about whether this is a good idea to the issue, not this PR.

**Comment by adriangb** [comment]
It seems like improving docs around subclassing `types.GenericAlias` as well as maybe improving some of the ergonomics of subclassing it is the path forward here, so I'm going to close this PR at least for now. Discussion can continue in the issue.

**Pull request closed without merging**